File: src/test/resources/de/hhu/stups/codegenerator/BBigInteger.cpp
#include <iostream>
#include <string>
#include "BBoolean.cpp"
#include "BigInt.hpp"

#ifndef BBIGINTEGER_H
#define BBIGINTEGER_H

using namespace std;

class BBigInteger : public BObject {

    /*
	@Override
	public boolean equals(Object obj) {
		if (== obj)
			return true;
		if (obj == null)
			return false;
		if (obj instanceof java.lang.Number) {
			return compareTo((java.lang.Number) obj) == 0;
		}
		// assert getClass() != obj.getClass()
		return false;
	}*/

	private:
	    BigInt value;

	/*@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((value == null) ? 0 : value.hashCode());
		return result;
	}*/

    public:
        /*BInteger(string val) {
            value = new java.math.BigInteger(val);
        }*/

        BBigInteger(const string& val) {
            value = BigInt(val);
        }

        BBigInteger(const BigInt& val) {
            value = val;
        }

        int intValue() const {
            return value.to_int();
        }

        BBigInteger(){};

        BBigInteger(const BBigInteger& val) {
            value = val.value;
        }

        BBoolean lessEqual(const BBigInteger& o) {
            return value <= o.value;
        }

        BBoolean greaterEqual(const BBigInteger& o) {
            return value >= o.value;
        }

        BBoolean less(const BBigInteger& o) {
            return value < o.value;
        }

        BBoolean greater(const BBigInteger& o) {
            return value > o.value;
        }

        BBoolean equal(const BBigInteger& o) {
            return value == o.value;
        }

        BBoolean unequal(const BBigInteger& o) {
            return value != o.value;
        }

        /*int compareTo(java.lang.Number o) {
            java.math.BigInteger oi;
            if (o == null) {
                throw new NullPointerException();
            }
            if (getClass() != o.getClass()) {
                oi = new java.math.BigInteger(java.lang.String.valueOf(o.longValue()));
            } else {
                BInteger oo = (BInteger) o;
                oi = oo.value;
            }
            return value.compareTo(oi);
        }*/

        /*long longValue() {
            return value.longValue();
        }

        float floatValue() {
            return value.floatValue();
        }

        double doubleValue() {
            return value.doubleValue();
        }*/

        BBigInteger plus(const BBigInteger& o) {
            return value + o.value;
        }

        /*java.math.BigInteger asBigInteger() {
            return value;
        }

        java.lang.String toString() {
            return value.toString();
        }*/

        BBigInteger minus(const BBigInteger& o) {
            return value - o.value;
        }

        BBigInteger multiply(const BBigInteger& o) {
            return value * o.value;
        }

        BBigInteger divide(const BBigInteger& o) {
            return value / o.value;
        }

        BBigInteger modulo(const BBigInteger& o) {
            return value % o.value;
        }

        /*BInteger or(BInteger o) {
            return new BInteger(value.or(o.value));
        }

        BInteger and(BInteger o) {
            return new BInteger(value.and(o.value));
        }

        BInteger xor(BInteger o) {
            return new BInteger(value.xor(o.value));
        }*/

        BBigInteger next() {
            return value + 1;
        }

        BBigInteger previous() {
            return value - 1;
        }

        /*BInteger leftShift(BInteger o) {
            return new BInteger(value.shiftLeft(o.intValue()));
        }

        BInteger rightShift(BInteger o) {
            return new BInteger(value.shiftRight(o.intValue()));
        }

        boolean isCase(BInteger o) {
            return equals(o);
        }*/

        BBigInteger negative() {
            return -value;
        }

        BBigInteger positive() {
            return *this;
        }

        /*BInteger operator=(BInteger other) {
            this->value = other.value;
            return *this;
        }*/

        friend bool operator !=(const BBigInteger& o1, const BBigInteger& o2) {
            return o1.value != o2.value;
        }

        friend bool operator ==(const BBigInteger& o1, const BBigInteger& o2) {
            return o1.value == o2.value;
        }

        void operator =(const BBigInteger& other) {
            value = other.value;
        }

        int hashCode() const {
            int prime = 31;
            int result = 1;
            result = prime * result + value.to_int();
            return result;
        }

};
#endif

File: btypes/src/main/java/de/hhu/stups/btypes/BBigInteger.java
package de.hhu.stups.btypes;


import clojure.lang.BigInt;
import clojure.lang.RT;
import clojure.lang.Var;

import java.math.BigInteger;

public class BBigInteger extends java.lang.Number implements Comparable<BBigInteger>, BObject {

	private static final Var PLUS = RT.var("clojure.core", "+");

	private static final Var MINUS = RT.var("clojure.core", "-");

	private static final Var MULTIPLY = RT.var("clojure.core", "*");

	private static final Var DIVIDE = RT.var("clojure.core", "quot");

	private static final Var MODULO = RT.var("clojure.core", "mod");

	private static final Var COMPARE = RT.var("clojure.core", "compare");

	public static final Var INC = RT.var("clojure.core", "inc");

	private static final Var DEC = RT.var("clojure.core", "dec");

	private static final Var POWER = RT.var("clojure.core", "^");

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (obj instanceof BBigInteger) {
			//TODO: other numbers
			return this.compareTo((BBigInteger) obj) == 0;
		}
		// assert getClass() != obj.getClass()
		return false;
	}

	private static final long serialVersionUID = -6484548796859331267L;
	private java.lang.Number value;

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((value == null) ? 0 : value.hashCode());
		return result;
	}

	public BBigInteger(String value) {
		try {
			this.value = Long.parseLong(value);
		} catch (NumberFormatException e) {
			this.value = BigInt.fromBigInteger(new java.math.BigInteger(value));
		}
	}

	public BBigInteger(java.lang.Number value) {
		this.value = value;
	}

	public int compareTo(BBigInteger o) {
		BBigInteger other = (BBigInteger) o;
		return (int) COMPARE.invoke(this.value, other.value);
	}

	public BBoolean lessEqual(BBigInteger o) {
		return new BBoolean(compareTo(o) <= 0);
	}


	public BBoolean greaterEqual(BBigInteger o) {
		return new BBoolean(compareTo(o) >= 0);
	}

	public BigInteger asBigInteger() {
		return new java.math.BigInteger(value.toString());
	}

	public BBoolean less(BBigInteger o) {
		return new BBoolean(compareTo(o) < 0);
	}

	public BBoolean greater(BBigInteger o) {
		return new BBoolean(compareTo(o) > 0);
	}

	public BBoolean equal(BBigInteger o) {
		return new BBoolean(compareTo(o) == 0);
	}

	public BBoolean unequal(BBigInteger o) {
		return new BBoolean(compareTo(o) != 0);
	}

	@Override
	public int intValue() {
		return this.value.intValue();
	}

	@Override
	public long longValue() {
		return this.value.longValue();
	}

	@Override
	public float floatValue() {
		return this.value.floatValue();
	}

	@Override
	public double doubleValue() {
		return this.value.doubleValue();
	}

	public BBigInteger plus(BBigInteger o) {
		return new BBigInteger((java.lang.Number) PLUS.invoke(this.value, o.value));
	}

	public java.lang.String toString() {
		return this.value.toString();
	}

	public BBigInteger minus(BBigInteger o) {
		return new BBigInteger((java.lang.Number) MINUS.invoke(this.value, o.value));
	}

	public BBigInteger multiply(BBigInteger o) {
		return new BBigInteger((java.lang.Number) MULTIPLY.invoke(this.value, o.value));
	}

	public BBigInteger power(BBigInteger o) {
		return new BBigInteger((java.lang.Number) POWER.invoke(this.value, o.value));
	}

	public BBigInteger divide(BBigInteger o) {
		return new BBigInteger((java.lang.Number) DIVIDE.invoke(this.value, o.value));
	}

	public BBigInteger modulo(BBigInteger o) {
		return new BBigInteger((java.lang.Number) MODULO.invoke(this.value, o.value));
	}

	/*@Override
	public BBigInteger or(BBigInteger o) {
		return null;
	}

	@Override
	public BBigInteger and(BBigInteger o) {
		return null;
	}

	@Override
	public BBigInteger xor(BBigInteger o) {
		return null;
	}*/

	public BBigInteger next() {
		return new BBigInteger((java.lang.Number) INC.invoke(this.value));
	}

	public BBigInteger previous() {
		return new BBigInteger((java.lang.Number) DEC.invoke(this.value));
	}

	public boolean isCase(BBigInteger o) {
		return this.equals(o);
	}

	public BBigInteger negative() {
		return new BBigInteger((java.lang.Number) MINUS.invoke(this.value));
	}

	public BBigInteger positive() {
		return this;
	}

	public java.lang.Number getValue() {
		return value;
	}
}